pyClient/test/test_contract_base_mixer.py

```diff
@@ -0,0 +1,204 @@
+# residual_bits = 713623846352979940490457358497079434602616037, or in bits
+# 1-4:   00000000 00000000 00000000 00000000
+# 5-8:   00000000 00000000 00000000 00000000
+# 9-12:  00000000 00000000 00000000 00000000
+# 13-16: 00000000 00011111 11111111 11111111
+# 17-20: 11111111 11111111 11111111 11111111
+# 21-24: 11111111 11100000 00000000 00000000
+# 25-28: 00000000 00000000 00000000 00000000
+# 29-32: 00000000 00011100 00010100 11100101
+# This corresponds to
+# v_in  = "0xFFFFFFFFFFFFFFFF" = 2**PUBLIC_VALUE_LENGTH - 1
+# v_out = "0x0000000000000000" = 0
+# h_sig = "111" = 7
+# sn_0  = "000" = 0
+# sn_1  = "001" = 1
+# cm_0  = "010" = 2
+# cm_1  = "011" = 3
+# h_0   = "100" = 4
+# h_1   = "101" = 5
+# The values were set to be easily distinguishable.
+
+PRIMARY_INPUTS = [
+    0,  # root
+    1,  # sn_0
+    1,  # sn_1
+    2,  # cm_0
+    2,  # cm_1
+    3,  # h_sig
+    4,  # h_0
+    4,  # h_1
+    713623846352979940490457358497079434602616037] \
+        # pylint: disable=no-member,invalid-name
+
+RESIDUAL_BITS = [
+    2**PUBLIC_VALUE_LENGTH - 1,  # v_in
+    0,  # v_out
+    7,  # h_sig
+    0,  # sn_0
+    1,  # sn_1
+    2,  # cm_0
+    3,  # cm_1
+    4,  # h_0
+    5  # h_1
+    ]  # pylint: disable=no-member,invalid-name
+
+
+def test_assemble_nullifiers(mixer_instance: Any) -> int:
+    # Test retrieving nullifiers
+    print("--- testing ", "test_assemble_nullifiers")
+    for i in range(JS_INPUTS):
+        res = mixer_instance.functions.\
+            assemble_nullifier(i, PRIMARY_INPUTS).call()
+        val = int.from_bytes(res, byteorder="big")
+        # We need to recompute the expected value
+        # To do so, we load the variable's first FIELD_CAPACITY bits from
+        # `PRIMARY_INPUTS` and remove the padding
+        # (of size DIGEST_LENGTH-FIELD_CAPACITY)
+        # Before adding the value defined in `residual bits`
+        expected_val = PRIMARY_INPUTS[1+i]*2**(DIGEST_LENGTH-FIELD_CAPACITY) +\
+            RESIDUAL_BITS[3+i]
+        if val != expected_val:
+            print("ERROR: extracted wrong nullifier")
+            print("expected:", expected_val, i)
+            print("got:", val, i)
+            return 1
+    return 0
+
+
+def test_assemble_commitments(mixer_instance: Any) -> int:
+    # Test retrieving commitments
+    print("--- testing ", "test_assemble_commitments")
+    for i in range(JS_OUTPUTS):
+        res = mixer_instance.functions.\
+            assemble_commitment(i, PRIMARY_INPUTS).call()
+        val = int.from_bytes(res, byteorder="big")
+        # We need to recompute the expected value
+        # To do so, we load the variable's first FIELD_CAPACITY bits from
+        # `PRIMARY_INPUTS` and remove the padding
+        # (of size DIGEST_LENGTH-FIELD_CAPACITY)
+        # Before adding the value defined in `residual bits`
+        expected_val = PRIMARY_INPUTS[1+JS_INPUTS+i] *\
+            2**(DIGEST_LENGTH-FIELD_CAPACITY) +\
+            RESIDUAL_BITS[3+JS_INPUTS+i]
```

Comment:
I found these to be confusing, which is problematic since it should be effortless to reason about such a simple test.

Let's modify the test and the comments by:
1. Providing the inputs before packing (the expected values)
2. Writing the inputs after packing (basically as you did)
3. Use the values in 1. as expected values instead of recomposing the expected values from either the loop index or the packed values

Reply:
Ok!

Reply:
`PRIMARY_INPUTS` should be updated to `PACKED_PRIMARY_INPUTS`, and we need to detail from which unpacked primary_inputs we obtained them to see if we recover the same values at the end

Reply:
> and we need to detail from which unpacked primary_inputs we obtained them 

I am not sure if you want to differentiate the unpacked primary from one another (as follows) or if you are saying something else.
```
UNPACKED_SN = [
    8,
    9
]
```

Reply:
You can write them all in the same array, with a comment next to it to specify which input it is, e.g. `8, # sn_0`.


Btw, we should stick to `nullifier` instead of `serial number`

Reply:
Fixed.